packages/compartment-mapper/src/import-hook.js

```diff
@@ -62,30 +63,56 @@ const nodejsConventionSearchSuffixes = [
 /**
  * @param {ReadFn|ReadPowers} readPowers
  * @param {string} baseLocation
- * @param {Sources} sources
- * @param {Record<string, CompartmentDescriptor>} compartmentDescriptors
- * @param {Record<string, any>} exitModules
- * @param {HashFn=} computeSha512
- * @param {Array<string>} searchSuffixes - Suffixes to search if the unmodified specifier is not found.
+ * @param {object} args
+ * @param {string} args.entryCompartmentName
+ * @param {string} args.entryModuleSpecifier
+ * @param {Sources} [args.sources]
+ * @param {Record<string, CompartmentDescriptor>} [args.compartments]
+ * @param {Record<string, any>} [args.exitModules]
+ * @param {HashFn} [args.computeSha512]
+ * @param {Array<string>} [args.searchSuffixes] - Suffixes to search if the
+ * unmodified specifier is not found.
  * Pass [] to emulate Node.js’s strict behavior.
  * The default handles Node.js’s CommonJS behavior.
- * Unlike Node.js, the Compartment Mapper lifts CommonJS up, more like a bundler,
- * and does not attempt to vary the behavior of resolution depending on the
- * language of the importing module.
+ * Unlike Node.js, the Compartment Mapper lifts CommonJS up, more like a
+ * bundler, and does not attempt to vary the behavior of resolution depending
+ * on the language of the importing module.
  * @returns {ImportHookMaker}
  */
 export const makeImportHookMaker = (
   readPowers,
   baseLocation,
-  sources = Object.create(null),
-  compartmentDescriptors = Object.create(null),
-  exitModules = Object.create(null),
-  computeSha512 = undefined,
-  searchSuffixes = nodejsConventionSearchSuffixes,
+  {
+    sources = Object.create(null),
+    compartments: compartmentDescriptors = Object.create(null),
+    exitModules = Object.create(null),
+    computeSha512 = undefined,
+    searchSuffixes = nodejsConventionSearchSuffixes,
+    entryCompartmentName,
+    entryModuleSpecifier,
+  },
 ) => {
-  // Set of specifiers for modules whose parser is not using heuristics to determine imports
-  const strictlyRequired = new Set();
-  // per-assembly:
+  // Set of specifiers for modules (scoped to compartment) whose parser is not
+  // using heuristics to determine imports.
+  /** @type {Map<string, Set<string>>} compartment name ->* module specifier */
+  const strictlyRequired = new Map([
+    [entryCompartmentName, new Set([entryModuleSpecifier])],
+  ]);
```

Comment:
Why does strictlyRequired need to depend on a compartment? If a module is strictly required at least once in the entire compartment-map, it should not be deferred. 
Why was putting the entryModuleSpecifier in the set not enough?
Why would it be ok to defer a module in one compartment while it's strictlyRequired in another compartment?

I expect to find out when I reach the new tests, but the only reason I can think of right now is if the specifier we're using as a set entry could point to different things in different compartments, in which case we probably should look for a different thing to identify the package with?

On the other hand, whether this is per compartment or not doesn't affect the end result - the one compatment where ESM import makes a specifier strictly required will be enough to cause an error to be thrown instead of the bundle/archive being created. 
It might take a few more operations to get there though. With a flat set if the first attempt to get the strictly required module was a cjs require, it would also throw AFAIR.

I'll get back to this comment when I finish. This is how far I got for now. Back in a few hours.

Reply:
Module specifiers are scoped to compartment. For example, `./index.js` refers to a different module in compartment A and compartment B.

Reply:
Yeah, got it. So we could fall back to path, but that would be more difficult and less performant than this.